File: inverted-index/InvertedIndex.scala
import com.twitter.scalding._

/** Build an inverted index!
  * 
  * An inverted index maps words to the documents
  * they appear in.
  * 
  * Run this in local mode with:
  *   scald.rb --local InvertedIndex.scala
  */
class InvertedIndex(args : Args) extends Job(args) {

  val tweets = Tsv("tweets.tsv", ('id, 'text))

  val wordToTweets =
    tweets
      .flatMap(('id, 'text) -> ('word, 'tweetId)) { 
        fields : (Long, String) => 
        val (tweetId, text) = fields
        text.split("\\s+").map { word => (word, tweetId) }
      }

  val invertedIndex =
    wordToTweets.groupBy('word) {  _.toList[Long]('tweetId -> 'tweetIds) }
    
  invertedIndex.write(Tsv("output.tsv"))
}

File: inverted-index/inverted_index.clj
(defmapcatop tokenize [text]
  (seq (.split text "\\s+")))

  (defaggregateop conj-words
    ([] [])
    ([word-list word]
      (conj word-list word))
    ([word-list]
      [word-list]))


(defn inverted-index [input-filename]
  (let [input (hfs-textline input-filename)]
    (<- [?word ?tweet-ids]
        (input ?textline)
        (tokenize ?textline :> ?word)
        (conj-words ?tweet-ids))))

(?- (stdout) (word-count "tweets.tsv"))